File: my_cpp_package/src/action_server_node.cpp
#include <chrono>
#include <thread>

#include "rclcpp/rclcpp.hpp"
#include "rclcpp_action/rclcpp_action.hpp"
#include "my_msgs/action/move_robot.hpp"

/**
 * @author: Mehmet Kahraman
 * @date: 09.10.2023
 * @about: Action server node
 **/


using namespace std::placeholders;
using MyActionMoveRobot = my_msgs::action::MoveRobot;
using GoalHandleMoveRobot = rclcpp_action::ServerGoalHandle<MyActionMoveRobot>;

class MyRosClass : public rclcpp::Node
{   

    public:
        
        MyRosClass() : Node("action_server_node")
        {   
            std::string action_topic = "/my_action_topic";
            
            action_server = rclcpp_action::create_server<MyActionMoveRobot>(
                this, 
                action_topic,
                std::bind(&MyRosClass::handle_goal, this, _1, _2),
                std::bind(&MyRosClass::handle_cancel, this, _1),
                std::bind(&MyRosClass::handle_accepted, this, _1));

            rclcpp::sleep_for(std::chrono::seconds(1));
            RCLCPP_INFO(this->get_logger(), "Action server node is ready.");
        }

    private:

        std::string service_topic;
        rclcpp_action::Server<MyActionMoveRobot>::SharedPtr action_server;

        rclcpp_action::GoalResponse handle_goal(
            const rclcpp_action::GoalUUID & uuid,
            std::shared_ptr<const MyActionMoveRobot::Goal> goal)
        {
            RCLCPP_INFO(this->get_logger(), "Received goal request.");
            (void)uuid;
            return rclcpp_action::GoalResponse::ACCEPT_AND_EXECUTE;
        }

        rclcpp_action::CancelResponse handle_cancel(
            const std::shared_ptr<GoalHandleMoveRobot> goal_handle)
        {
            RCLCPP_INFO(this->get_logger(), "Received request to cancel goal");
            (void)goal_handle;
            return rclcpp_action::CancelResponse::ACCEPT;
        }

        void handle_accepted(const std::shared_ptr<GoalHandleMoveRobot> goal_handle)
        {
            std::thread{std::bind(&MyRosClass::execute_action, this, _1), goal_handle}.detach();
        }

        void execute_action(const std::shared_ptr<GoalHandleMoveRobot> goal_handle)
        {
            RCLCPP_INFO(this->get_logger(), "Action started. Executing goal...");

            const auto goal = goal_handle->get_goal();
            auto feedback = std::make_shared<MyActionMoveRobot::Feedback>();
            auto result = std::make_shared<MyActionMoveRobot::Result>();

            rclcpp::Time start_time = this->now();
            rclcpp::Time current_time = this->now();
            rclcpp::Duration elapsed_time = current_time - start_time;
            double seconds = 0.0;

            int progress = 0;
            int loop_hz = 1;
            rclcpp::Rate loop_rate(loop_hz);

            while (rclcpp::ok())
            {      
                if (goal_handle->is_canceling() == true) 
                {
                    result->finish_success = false;
                    goal_handle->canceled(result);
                    RCLCPP_INFO(this->get_logger(), "Goal canceled.");
                    return;  //break
                }

                feedback->progress = progress;
                goal_handle->publish_feedback(feedback);
                RCLCPP_INFO(this->get_logger(), "Action progress: %d percent.", progress);

                current_time = this->now();
                elapsed_time = current_time - start_time;
                seconds = elapsed_time.seconds();
                RCLCPP_INFO(this->get_logger(), "Elapsed time: %f seconds.", seconds);

                if (progress == 100) {
                    RCLCPP_INFO(this->get_logger(), "Progress done.");
                    break;
                }

                progress = progress + 10;
                loop_rate.sleep();
            }

            if (rclcpp::ok() && progress == 100) {
                result->finish_success = true;
                goal_handle->succeed(result);
                RCLCPP_INFO(this->get_logger(), "Action goal successfully done.");
            }
        }
    

}; // class


int main(int argc, char** argv)
{
    rclcpp::init(argc, argv);
    auto action_server_node = std::make_shared<MyRosClass>();
    rclcpp::spin(action_server_node);

    rclcpp::shutdown();
    return 0;
}

File: my_py_package/my_py_package/action_server_node.py
#!/usr/bin/env python3

import time
import rclpy
from rclpy.node import Node
from rclpy.action import ActionServer, CancelResponse, GoalResponse
from my_msgs.action import MoveRobot

"""
  @author: Mehmet Kahraman
  @date: 21.02.2024
  @about: Action server node
"""

class MyROSClass(Node):

    def __init__(self):

        super().__init__('action_server_node')

        self.service_topic = "/my_action_topic"
    
        self.action_server = ActionServer(self, 
            MoveRobot, 
            self.service_topic, 
            execute_callback=self.execute_action,
            goal_callback=self.handle_goal,
            cancel_callback=self.handle_cancel
        )
        
        time.sleep(1)
        self.get_logger().info("Action server is ready.")
    

    def handle_goal(self, goal_request):
        self.get_logger().info('Received goal request.')
        return GoalResponse.ACCEPT

    def handle_cancel(self, goal_handle):
        self.get_logger().warn('Received request to cancel goal')
        return CancelResponse.ACCEPT

    def execute_action(self, goal_handle):
        self.get_logger().info('Action started. Executing goal...')

        feedback = MoveRobot.Feedback()
        result = MoveRobot.Result()

        start_time = self.get_clock().now()
        current_time = self.get_clock().now()
        elapsed_time = current_time - start_time
        seconds = 0.0

        progress = 0
        loop_hz = 1
        loop_rate = 1/loop_hz

        while rclpy.ok():
            
            if goal_handle.is_cancel_requested == True:
                result.finish_success = False
                goal_handle.canceled(result)
                self.get_logger().warn('Goal canceled.')
                return  # break
            
            feedback.progress = progress
            goal_handle.publish_feedback(feedback)
            self.get_logger().info(f'Action progress: {progress} percent.')

            current_time = self.get_clock().now()
            elapsed_time = current_time - start_time
            seconds = elapsed_time.nanoseconds / 1e9
            self.get_logger().info(f'Elapsed time: {seconds} seconds.')

            if progress == 100:
                self.get_logger().info('Progress done.')
                break

            progress = progress + 10
            time.sleep(loop_rate)

        if (rclpy.ok() == True) and (progress == 100):
            result.finish_success = True
            goal_handle.succeed()
            self.get_logger().info('Action goal successfully done.')

        return result


def main():
    rclpy.init()
    node = MyROSClass()
    rclpy.spin(node)

    Node.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()
